### PR TITLE
Various Fixes to SetPartial + Set on abstract object

### PR DESCRIPTION
**Pull request by sebmarkbage**
This is the equivalent of #1680 but for assignments.

It's pretty much the same changes, except this one emits assignments.

I also had to implement assignment to topVal.

### CODE REVIEW

File: src/methods/properties.js [hunk 1]
```diff
@@ -995,6 +995,11 @@ export class PropertiesImplementation {
 
     // 6. Else if IsPropertyReference(V) is true, then
     if (Environment.IsPropertyReference(realm, V)) {
+      if (base instanceof AbstractValue) {
+        // Ensure that abstract values are coerced to objects. This might yield
+        // an operation that might throw.
+        base = To.ToObject(realm, base);
+      }
```
Comment (sebmarkbage) [inline src/methods/properties.js]: This ensures that if we throw a fatal if we're not in pure mode already. It is the equivalent for what _dereference does. https://github.com/facebook/prepack/blob/88fbdcc47b15a38ac4369b7eeb893e2d62eedba2/src/methods/environment.js#L137-L141

File: src/values/AbstractObjectValue.js [hunk 2]
```diff
@@ -628,18 +628,70 @@ export default class AbstractObjectValue extends AbstractValue {
...
+        // The emitted assignment might throw at runtime but if it does, that
+        // is handled by evaluateWithPossibleThrowCompletion. Anything that
+        // happens after this, can assume we didn't throw and therefore,
+        // we return true here.
+        return true;
+      }
+      let error = new CompilerDiagnostic(
+        "property access on unknown object",
+        this.$Realm.currentLocation,
+        "PP0031",
```
Comment (hermanventer) [inline src/values/AbstractObjectValue.js]: It is not clear to me why PP0031 needs to exist, given that we already have PP0012.

File: src/values/AbstractObjectValue.js [hunk 3]
```diff
@@ -668,26 +720,66 @@ export default class AbstractObjectValue extends AbstractValue {
 
   $SetPartial(P: AbstractValue | PropertyKeyValue, V: Value, Receiver: Value): boolean {
     if (!(P instanceof AbstractValue)) return this.$Set(P, V, Receiver);
-    invariant(this === Receiver, "TODO #1021");
     if (this.values.isTop()) {
-      AbstractValue.reportIntrospectionError(this);
+      if (this.$Realm.isInPureScope()) {
```
Comment (hermanventer) [inline src/values/AbstractObjectValue.js]: This seems to be a duplicate of the code in line 632. You can avoid the duplication by moving line 722 below this if.

File: src/values/ObjectValue.js [hunk 4]
```diff
@@ -782,6 +784,77 @@ export default class ObjectValue extends ConcreteValue {
...
+        // The emitted assignment might throw at runtime but if it does, that
+        // is handled by evaluateWithPossibleThrowCompletion. Anything that
+        // happens after this, can assume we didn't throw and therefore,
+        // we return true here.
+        return true;
+      } else {
+        let error = new CompilerDiagnostic(
+          "unknown property access might need to invoke a setter",
+          this.$Realm.currentLocation,
+          "PP0030",
```
Comment (hermanventer) [inline src/values/ObjectValue.js]: Why is this a better error message than PP0011?

File: src/values/ObjectValue.js [hunk 5]
```diff
@@ -782,6 +784,77 @@ export default class ObjectValue extends ConcreteValue {
...
+          "RecoverableError"
+        );
+        if (this.$Realm.handleError(error) !== "Recover") {
+          throw new FatalError();
+        }
+      }
+    }
+
+    // We should never consult the prototype chain for unknown properties.
+    // If it was simple, it would've been an assignment to the receiver.
```
Comment (hermanventer) [inline src/values/ObjectValue.js]: If it were simple

File: src/values/ObjectValue.js [hunk 6]
```diff
@@ -782,6 +784,77 @@ export default class ObjectValue extends ConcreteValue {
...
+    // The only case the Receiver isn't this, if this was a ToObject
```
Comment (hermanventer) [inline src/values/ObjectValue.js]: , is if

File: src/values/AbstractObjectValue.js (hunk 3 above)
Comment (sebmarkbage) [reply src/values/AbstractObjectValue.js]: Not all callsites go through $SetPartial though. I guess ideally they should tho, right?

Comment (hermanventer) [reply src/values/AbstractObjectValue.js]: Meanwhile you could call $SetPartial from $Set in this particular case.

**Pull request closed without merging**